### DISCUSSION TITLE
How to get NumLock state in Linux

### DISCUSSION THREAD

**Discussion by PermanentNovice** [AvaloniaUI/Avalonia#15173] [Q&A]
Someone stated [here](https://github.com/AvaloniaUI/Avalonia/issues/2422) that Avalonia won't create a global API for determining the state of the caps lock key. I need to get the state of the numlock key and found what I need using DllImport to call the WinApi in just two lines of code.

Can I do something similar in Linux(Gnome)? I know nothing about calling native libraries so please direct me where I can simply copy the code.

**Comment by wsficke** [comment]
Depending on your GUI, you would use the corresponding display server ABI.  This could involve PInvoke into Wayland or X11 code, for example.  Review unmanaged extensibility in your environment and then create PInvoke bindings for .NET interop.

> I know nothing about calling native libraries

It is worthwhile to research and learn how unmanaged method calling works.  Start here:

https://en.wikipedia.org/wiki/X86_calling_conventions

**Comment by PermanentNovice** [comment]
For now, it will be for X11. In the original post, I mentioned I would need code I can simply copy. Can anyone expand on this? From the answer, I assume it will not be a simple two line function as it is for Win32.

**Comment by wsficke** [comment]
https://stackoverflow.com/questions/9009775/using-getkeystatevk-capital-1-in-linux
https://learn.microsoft.com/en-us/dotnet/standard/native-interop/pinvoke-source-generation

There's not much discussion of managed access on the search engines, but plenty of C++ discussions on this subject.  Looks like your target invocation would be XkbGetIndicatorState(a, b, c), which is declared in X11/XKBlib.h.  So you use P/Invoke; extern static method with DllImportAttribute to bind the extern in XKB.lib.  Your static class of static externs decorated with DllImport attributes works for managed code like how the header file works for C++; it provides link-time information to the compiler and encapsulates dynamic loading and marshalling of parameters and returns.  

We do this on Windows too, for accessing Win32 ABIs that don't have built-in managed access.  It's a core technique for platform targeting.  So, to your request for "code I can simply copy"; I say _no_.  If you are going to work in this space it is best to understand what you are doing; so you can address gaps in the work of others.  Otherwise you will have no end of frustration.  Unmanaged systems interoperation requires extreme attention to detail.

These problems come up reasonably often and people end up making nuget packages that wrap up the interop types.  This one (https://github.com/ajnewlands/X11.Net) is missing your API and seems abandoned.  If you want to do a solid to those who come after you, fork it and extend it with your P/Invoke code.  Then the community will have a code to simply copy.

**Comment by PermanentNovice** [comment]
Thanks. I will post on StackOverflow and hope someone can turn that into c# code for me. Do I need to mention Avalonia in the question? Or is it just a matter of calling an X11 library from C# and not related to Avalonia itself?

**Reply by timunie** [reply]
worth to mention Avalonia is used, but it's not required. It should be the exact same code for a plain console App for sure.
